crates/sui-graphql-rpc/src/types/cursor.rs

```diff
@@ -0,0 +1,357 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{fmt, ops::Deref};
+
+use async_graphql::{
+    connection::{CursorType, OpaqueCursor},
+    *,
+};
+use serde::{de::DeserializeOwned, Serialize};
+
+use crate::{config::ServiceConfig, error::Error};
+
+/// Wrap the `OpaqueCursor` type to implement Scalar for it.
+pub(crate) struct Cursor<C>(OpaqueCursor<C>);
+
+/// Connection field parameters parsed into a single type that encodes
+#[derive(Debug)]
+pub(crate) struct Page<C> {
+    after: Option<Cursor<C>>,
+    before: Option<Cursor<C>>,
+    limit: u64,
+    side: Side,
+}
+
+/// Whether the page is extracted from the beginning or the end of the range bounded by the cursors.
+#[derive(PartialEq, Eq, Debug)]
+enum Side {
+    Prefix,
+    Suffix,
+}
+
+impl<C> Cursor<C> {
+    pub(crate) fn new(cursor: C) -> Self {
+        Cursor(OpaqueCursor(cursor))
+    }
+}
+
+impl<C> Page<C> {
+    pub(crate) fn from_params(
+        config: &ServiceConfig,
+        first: Option<u64>,
+        after: Option<Cursor<C>>,
+        last: Option<u64>,
+        before: Option<Cursor<C>>,
+    ) -> Result<Self> {
+        let limits = &config.limits;
+        let page = match (first, after, last, before) {
+            (Some(_), _, Some(_), _) => return Err(Error::CursorNoFirstLast.extend()),
+
+            (limit, after, None, before) => Page {
+                after,
+                before,
+                limit: limit.unwrap_or(limits.default_page_size),
+                side: Side::Prefix,
+            },
+
+            (None, after, Some(limit), before) => Page {
+                after,
+                before,
+                limit,
+                side: Side::Suffix,
+            },
+        };
+
+        if page.limit > limits.max_page_size {
```

Comment:
I am confused about this condition, why would this have to throw an Error, can't wrap my mind around it? Requesting `last = 60` would throw an error, right?

Reply:
Yes, if you request a larger page than the service supports as a max page size, it will cause an error.  The default max page size is `50` so setting `first` or `last` to `60` would indeed cause it to error.

I'm assuming an implied alternative which is to cap the limit to the max page size? I didn't want to do that because it breaks an assumption that the page size is different from requested only when you are on the last page.  I would also rather push clients towards a pattern where they query for the max page size ahead of time and then adhere to it than let them pretend like it doesn't exist.

Reply:
After reading your comment and thinking more about it, it makes sense to err rather than "deceive" the user.